pubsignals/common.go

```diff
@@ -0,0 +1,248 @@
+)
+
+// ParseCredentialSubject parse credential subject and return array of property queries
+func ParseCredentialSubject(_ context.Context, credentialSubject any) (out []PropertyQuery, err error) {
+	if credentialSubject == nil {
+		return []PropertyQuery{
+			{
+				Operator:  circuits.EQ,
+				FieldName: "",
+			},
+		}, nil
+	}
+
+	out = []PropertyQuery{}
+
+	jsonObject, ok := credentialSubject.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("Failed to convert credential subject to JSONObject")
+	}
+
+	entries := getObjectEntries(jsonObject)
+	if len(entries) == 0 {
+		return nil, errors.New("query must have at least 1 predicate")
+	}
+	for fieldName, fieldReq := range entries {
+		fieldReqEntries := getObjectEntries(fieldReq.(map[string]interface{}))
+		isSelectiveDisclosure := len(fieldReqEntries) == 0
+
+		if isSelectiveDisclosure {
+			out = append(out, PropertyQuery{Operator: circuits.SD, FieldName: fieldName})
+			continue
+		}
+
+		for operatorName, operatorValue := range fieldReqEntries {
+			if _, exists := circuits.QueryOperators[operatorName]; !exists {
+				return nil, errors.New("operator is not supported by lib")
+			}
+			operator := circuits.QueryOperators[operatorName]
+			out = append(out, PropertyQuery{Operator: operator, FieldName: fieldName, OperatorValue: operatorValue})
+		}
+	}
+
+	return out, nil
+}
+
+func getObjectEntries(obj map[string]interface{}) map[string]interface{} {
+	entries := make(map[string]interface{})
+	for k, v := range obj {
+		entries[k] = v
+	}
+	return entries
+}
+
+// ParseQueryMetadata parse property query and return query metadata
+func ParseQueryMetadata(ctx context.Context, propertyQuery PropertyQuery, ldContextJSON, credentialType string, options merklize.Options) (query *QueryMetadata, err error) {
+	datatype, err := options.TypeFromContext([]byte(ldContextJSON), fmt.Sprintf("%s.%s", credentialType, propertyQuery.FieldName))
+	if err != nil {
+		return nil, err
+	}
+
+	query = &QueryMetadata{
+		PropertyQuery:   propertyQuery,
+		SlotIndex:       0,
+		MerklizedSchema: false,
+		Datatype:        datatype,
+		ClaimPathKey:    big.NewInt(0),
+		Values:          []*big.Int{},
+		Path:            &merklize.Path{},
+	}
+
+	var ctxObj map[string]interface{}
+	err = json.Unmarshal([]byte(ldContextJSON), &ctxObj)
+	if err != nil {
+		return nil, err
+	}
+
+	ldCtx, err := ld.NewContext(nil, nil).Parse(ctxObj["@context"])
+	if err != nil {
+		return nil, err
+	}
+
+	serAttr, err := getSerializationAttrFromParsedContext(ldCtx, credentialType)
+	if err != nil {
+		return nil, err
+	}
+	if serAttr == "" {
+		query.MerklizedSchema = true
+	}
+
+	if !query.MerklizedSchema {
+		query.SlotIndex, err = parser.Parser{}.GetFieldSlotIndex(propertyQuery.FieldName, credentialType, []byte(ldContextJSON))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		path := merklize.Path{}
+		if query.FieldName != "" {
```

Comment:
if FieldName == "" , path should be a path to credentialSubject field

Reply:
fixed